File: frontend/lib/relationshipAnalysis/library/relationshipDecisionLibrary.ts
/**
 * Relationship Decision Library — backward-compatible view over Model Ontology.
 * SSOT: ontology/modelOntology.ts
 */

import type { DecisionOptionId } from "../types/relationship";
import { familyDecisionLogic, MODEL_ONTOLOGY, type ModelFamilyId } from "../ontology/modelOntology";
import { routeModel, type ModelRouteResult } from "../ontology/modelRouter";
import {
  buildCardFromLibraryModel,
  instantiateModelCard,
  validateCardOntology,
  constrainLlmFillToOntology,
  ontologyTemplateForPrompt,
} from "../ontology/cardTemplateSystem";

export type RelationshipLibraryModelId = Exclude<ModelFamilyId, "generic">;

export type RelationshipLibraryModel = {
  id: RelationshipLibraryModelId;
  phaseOrder: 0 | 1 | 2;
  nextPhase?: RelationshipLibraryModelId;
  title: string;
  problemType: string;
  modelSummary: string;
  signalModel: { keyPositiveSignals: string[]; keyNegativeSignals: string[] };
  triggerConditions: string[];
  decisionLogic: string;
  decisionLogicByOption: Record<DecisionOptionId, string>;
  riskModel: { coreRisks: string[]; misjudgmentSources: string[] };
  actionTemplate: string[];
  reusabilityTags: string[];
};

function toLibraryView(id: RelationshipLibraryModelId): RelationshipLibraryModel {
  const family = MODEL_ONTOLOGY[id];
  const branches = family.canonicalStructure.decisionModel.fixedBranches;
  return {
    id,
    phaseOrder: family.phaseOrder ?? 0,
    nextPhase: family.nextPhase as RelationshipLibraryModelId | undefined,
    title: family.title,
    problemType: family.problemType,
    modelSummary: family.modelSummary,
    signalModel: family.template.signalModel,
    triggerConditions: family.template.triggerConditions,
    decisionLogic: familyDecisionLogic(family),
    decisionLogicByOption: branches,
    riskModel: family.template.riskModel,
    actionTemplate: family.template.actionTemplate,
    reusabilityTags: [...family.template.reusabilityTags],
  };
}

export const RELATIONSHIP_LIBRARY_PHASE_ORDER: RelationshipLibraryModelId[] = [
  "ambiguous_phase",
  "cold_phase",
  "breakdown_phase",
];

export const RELATIONSHIP_DECISION_LIBRARY: Record<
  RelationshipLibraryModelId,
  RelationshipLibraryModel
> = {
  ambiguous_phase: toLibraryView("ambiguous_phase"),
  cold_phase: toLibraryView("cold_phase"),
  breakdown_phase: toLibraryView("breakdown_phase"),
};

export { routeModel, routeModel as routeRelationshipModel, libraryModelLabel } from "../ontology/modelRouter";
export type { ModelRouteResult };

export {
  buildCardFromLibraryModel,
  instantiateModelCard,
  validateCardOntology,
  constrainLlmFillToOntology,
  ontologyTemplateForPrompt,
};

export {
  MODEL_ONTOLOGY,
  getModelFamily,
  familyDecisionLogic,
  type ModelFamily,
  type ModelFamilyId,
} from "../ontology/modelOntology";

File: src/gateway/services/relationship_decision_library.py
"""Backward-compatible view over Model Ontology."""

from __future__ import annotations

from typing import Any, Dict, Optional

from .relationship_model_ontology import (
    MODEL_ONTOLOGY,
    build_card_from_library_model,
    instantiate_model_card,
    library_template_for_prompt,
    ontology_template_for_prompt,
    route_model,
    route_relationship_model,
)

RELATIONSHIP_LIBRARY_PHASE_ORDER = ("ambiguous_phase", "cold_phase", "breakdown_phase")


def _to_library_view(family_id: str) -> Dict[str, Any]:
    family = MODEL_ONTOLOGY[family_id]
    branches = (family.get("canonicalStructure") or {}).get("decisionModel", {}).get("fixedBranches") or {}
    tpl = family.get("template") or {}
    return {
        "id": family_id,
        "phaseOrder": family.get("phaseOrder"),
        "nextPhase": family.get("nextPhase"),
        "title": family.get("title"),
        "problemType": family.get("problemType"),
        "modelSummary": family.get("modelSummary"),
        "signalModel": tpl.get("signalModel"),
        "triggerConditions": tpl.get("triggerConditions"),
        "decisionLogic": "\n".join(str(branches.get(k) or "") for k in ("A", "B", "C", "D")),
        "decisionLogicByOption": branches,
        "riskModel": tpl.get("riskModel"),
        "actionTemplate": tpl.get("actionTemplate"),
        "reusabilityTags": tpl.get("reusabilityTags"),
    }


RELATIONSHIP_DECISION_LIBRARY: Dict[str, Dict[str, Any]] = {
    family_id: _to_library_view(family_id) for family_id in RELATIONSHIP_LIBRARY_PHASE_ORDER
}
